R/downloader.R

```diff
@@ -198,7 +191,36 @@ downloadWithRetries <- function(url, ..., maxTries = 5L) {
     )
     if (success) break
   }
-  success
+  if (!success) {
+    stop("Download failed.")
+  }
+  return(TRUE)
+}
+
+canUseRenvDownload <- function() {
+  getOption("packrat.authenticated.downloads.use.renv", FALSE) &&
+    identical(secureDownloadMethod(), "curl")
+}
+
+renvDownload <- function(url, destfile, method = inferAppropriateDownloadMethod(url), type = NULL, ...) {
+  debug <- getOption("packrat.renv.debug", FALSE)
+  if (debug) {
+    options(renv.download.trace = TRUE) # TODO remove
+  }
+
+  # We temporarily set our user agent to "curl" so that Bitbucket will treat us
+  # like a command line and not a browser. Otherwise, if we make unauthorized
+  # requests to Bitbucket .tar.gz URLs, we get redirects instead of a 401.
+  renv_useragent_option <- options("renv.http.useragent" = "curl")
+  on.exit(options(renv_useragent_option), add = TRUE)
+
+  result <- with_envvar(
+    c(RENV_DOWNLOAD_METHOD = method),
+    renv$download(url = url, destfile = destfile, type = type)
+  )
+  if (identical(result, destfile)) {
```

Comment:
Is this return value meaningful? What should happen when `result` and `destfile` are not identical? Explain the contract of this function.

Maybe:
```r
return(identical(result, destfile))
```

Reply:
Update: I didn't use your suggestion — instead I used:

```r
  if (!file.exists(destfile)) {
    stop("No data received.", call. = FALSE)
  }
  return(TRUE)
```

Because `result` will always be identical to `destfile`, whereas if `renv`'s download method fails `destfile` will not exist.